src/box/memtx_hash.cc

```diff
@@ -286,9 +286,20 @@ memtx_hash_index_random(struct index *base, uint32_t rnd, struct tuple **result)
 
 	*result = NULL;
 	uint32_t k = light_index_random(hash_table, rnd);
-	if (k == light_index_end)
+	struct txn *txn = in_txn();
+	struct space *space = space_by_id(base->def->space_id);
+	if (k == light_index_end) {
+		/*
+		 * `light_index_end` is returned only in case the space is
+		 * empty.
+		 */
+		memtx_tx_track_full_scan(txn, space, base);
 		return 0;
+	}
 	*result = light_index_get(hash_table, k);
+	assert(*result != NULL);
+	assert(!base->def->key_def->is_multikey);
+	*result = memtx_tx_tuple_clarify(txn, space, *result, base, 0);
```

Comment:
What if we have found invisible tuple (clarified to NULL)?

Reply:
Discussed with @alyapunov: let's first check if the space is empty from the transaction's perspective, and if it's not repeatedly call `random` and clarify result, until we get a non-empty one.

We do not care about performance here, since all operations in context of
transaction management currently have O(number of dirty tuples)
complexity.